s3tests_boto3/functional/test_s3select.py

```diff
@@ -0,0 +1,410 @@
+
+    res_s3select_no_alias = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,"select int(_1),int(_2),int(_1)+int(_2) from stdin where (int(_1)+int(_2))>100 and (int(_1)+int(_2))<300;")  ).replace(",","")
+
+    assert res_s3select_alias == res_s3select_no_alias
+
+
+def test_alias_cyclic_refernce():
+
+    # purpose of test is to validate the s3select-engine is able to detect a cyclic reference to alias.
+    
+    csv_obj = create_random_csv_object(10000,10)
+
+    csv_obj_name = "csv_10000x10"
+    bucket_name = "test"
+    upload_csv_object(bucket_name,csv_obj_name,csv_obj)
+
+    res_s3select_alias = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,"select int(_1) as a1,int(_2) as a2, a1+a4 as a3, a5+a1 as a4, int(_3)+a3 as a5 from stdin;")  )
+
+    find_res = res_s3select_alias.find("number of calls exceed maximum size, probably a cyclic reference to alias");
+    
+    assert int(find_res) >= 0 
+
+def test_datetime():
+
+    # purpose of test is to validate date-time functionality is correct,
+    # by creating same groups with different functions (nested-calls) ,which later produce the same result 
+
+    csv_obj = create_csv_object_for_datetime(10000,1)
+
+    csv_obj_name = "csv_datetime_10000x10"
+    bucket_name = "test"
+
+    upload_csv_object(bucket_name,csv_obj_name,csv_obj)
+
+    res_s3select_date_time = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin where extract("year",timestamp(_1)) > 1950 and extract("year",timestamp(_1)) < 1960;')  )
+
+    res_s3select_substr = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin where int(substr(_1,1,4))>1950 and int(substr(_1,1,4))<1960;')  )
+
+    assert res_s3select_date_time == res_s3select_substr
+
+    res_s3select_date_time = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin where  datediff("month",timestamp(_1),dateadd("month",2,timestamp(_1)) ) == 2;')  )
+
+    res_s3select_count = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin;')  )
+
+    assert res_s3select_date_time == res_s3select_count
+
+    res_s3select_date_time = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin where datediff("year",timestamp(_1),dateadd("day", 366 ,timestamp(_1))) == 1 ;')  )
+
+    assert res_s3select_date_time == res_s3select_count
+
+    # validate that utcnow is integrate correctly with other date-time functions 
+    res_s3select_date_time_utcnow = remove_xml_tags_from_result(  run_s3select(bucket_name,csv_obj_name,'select count(0) from  stdin where datediff("hours",utcnow(),dateadd("day",1,utcnow())) == 24 ;')  )
+
+    assert res_s3select_date_time_utcnow == res_s3select_count
+
+def test_csv_parser():
+
+    # purpuse: test default csv values(, \n " \ ), return value may contain meta-char 
```

Comment:
could you please also verify:
* newline inside a quoted columns
* escaped escape charecter
could look here for a [list of ideas](https://github.com/maxogden/csv-spectrum/tree/master/csvs)

Reply:
escape-escaped is not implemented (i.e. successive escape characters ) 

newline (row-delimiter?) within quotes , if you meant "0xa" , in quotes , it's not supported (there is a test to validate that), since it may reduce performance (search for closing quote).

as for control-chars with escape (specific char sequence) , such as \t,\r,\n its not implemented.

"list of ideas" is ok, but need to priorities implementation efforts.

Reply:
we don't have to implement all aspect of CSV, but we must document exactly what we implemented. 
if we didn't implement the whole feature, our documentation cannot just add a reference to a spec (from the RFC, or AWS)

Reply:
in documentation there is a section(a table) describing different CSV parsing scenario's.
in case something is missing, i will add there.

Reply:
as mentioned, there is detailed table (in dedicated section) for CSV different cases.(more detailed than AWS)
ceph-s3-select is currently not compliant with AWS, we will probably need to change that once we integrate with data-analytic utility such as presto(for one example).

Reply:
ok. looked at the table again, i think it coverts everything (including our newline in quote behavior)